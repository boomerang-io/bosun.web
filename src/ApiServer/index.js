import { Server, Serializer, Model } from "miragejs";
import { inflections } from "inflected";
import uuid from "uuid/v4";
import { serviceUrl } from "Config/servicesConfig";
import * as fixtures from "./fixtures";

export function startApiServer({ environment = "test", timing = 0 } = {}) {
  inflections("en", function (inflect) {
    // Prevent pluralization bc our apis are weird
    inflect.irregular("policy", "policy");
    inflect.irregular("violation", "violation");
    inflect.irregular("template", "template");
    inflect.irregular("insights", "insights");
  });

  return new Server({
    environment,
    // Load in mock data
    fixtures,
    // Return the data as is, don't add a root key
    serializers: {
      application: Serializer.extend({
        root: false,
        embed: true,
      }),
    },
    // Register the data as a model so we can use the schema
    models: {
      templates: Model,
      policies: Model,
    },

    routes() {
      // Control how long the responses take to resolve
      this.timing = timing;

      // Allow unhandled requests on the current domain to pass through
      this.passthrough();

      this.get("/info", () => []);

      /**
       * Simple GET of static data
       */
      this.get(serviceUrl.getUserProfile(), (schema) => {
        return schema.db.profile[0];
      });
      this.get(serviceUrl.getCicdNavigation({}), (schema, request) => {
        return schema.db.cicdNavigation.map((item) => {
          if (typeof item.link === "string") {
            item.link = item.link.replace(":teamId", request.queryParams.teamId);
          }

          return item;
        });
      });

      this.get(serviceUrl.getPlatformNavigation(), (schema) => {
        return schema.db.navigation[0];
      });

      this.get(serviceUrl.getTeams(), (schema) => {
        return schema.db.teams;
      });

      this.get(serviceUrl.getPolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        return activePolicy;
      });

      this.get(serviceUrl.getValidateInfo({ policyId: ":policyId" }), (schema) => {
        return schema.db.validateInfo;
      });

      this.get(serviceUrl.getTemplates(), (schema) => {
        return schema.db.templates;
      });

      this.get(serviceUrl.getPolicyOverview(), (schema, request) => {
        let { teamId } = request.queryParams;
        return schema.db.policies.filter((policy) => policy.teamId === teamId);
      });

      this.get(serviceUrl.getViolationsOverview(), (schema) => {
        return schema.db.violations;
      });

      this.get(serviceUrl.getInsightsOverview(), (schema) => {
        return schema.db.insights;
      });

      this.post(serviceUrl.postCreatePolicy(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.policies.create({ ...body, id: uuid() });
      });

      this.post(serviceUrl.postCreatePolicyTemplate(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.templates.create({ ...body, id: uuid() });
      });

      this.patch(serviceUrl.patchUpdatePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activePolicy = schema.policies.find(policyId);
        activePolicy.update(body);
        return schema.policies.all();
      });

      this.patch(serviceUrl.patchUpdatePolicyTemplate({ templateId: ":templateId" }), (schema, request) => {
        let { templateId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTemplate = schema.templates.find(templateId);
        activeTemplate.update(body);
        return schema.templates.all();
      });

      this.delete(serviceUrl.deletePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        activePolicy.destroy({ id: policyId });
        return schema.policies.all();
      });
    },
  });
}
