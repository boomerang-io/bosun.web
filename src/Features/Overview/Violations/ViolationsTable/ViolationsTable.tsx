import React, { Component } from "react";
import moment from "moment";
import {
  Button,
  CodeSnippet,
  ComposedModal,
  DataTable,
  ModalFooter,
  ModalForm,
  ModalBody,
  Tag,
} from "@boomerang-io/carbon-addons-boomerang-react";
import copy from "copy-to-clipboard";
import { isAccessibleEvent } from "@boomerang-io/utils";
import { Violation } from "Types";
import styles from "./violationsTable.module.scss";

type Props = {
  violations: Violation[];
};

type State = any;

export class ViolationsTable extends Component<Props, State> {
  annotationsRef = React.createRef();

  state: { isModalOpen: boolean; selectedPolicyIndex: number | null } = {
    isModalOpen: false,
    selectedPolicyIndex: null,
  };
  headers = [
    {
      header: "Policy",
      key: "policyName",
    },
    {
      header: "Violations",
      key: "nbrViolations",
    },
    {
      header: "Failed Definition Types",
      key: "policyDefinitionTypes",
    },
    {
      header: "Reference ID",
      key: "referenceId",
    },
    {
      header: "Activity Date",
      key: "policyActivityCreatedDate",
    },
  ];

  subHeaders = [
    {
      header: "Metrics",
      key: "metric",
    },
    {
      header: "Message",
      key: "message",
    },
  ];

  renderCell = (rowIndex: any, cellIndex: any, value: any) => {
    const column = this.headers[cellIndex];
    switch (column.header) {
      case "Activity Date":
        return <time className={styles.tableTextarea}>{value ? moment(value).format("MMMM DD, YYYY") : "---"}</time>;
      case "Failed Definition Types":
        return <p className={styles.tableTextarea}>{value && value.length ? value.join(", ") : "---"}</p>;
      default:
        return <p className={styles.tableTextarea}>{value || "---"}</p>;
    }
  };

  renderSubRow = (row: any) => {
    const { violations } = this.props;
    const currentViolation = violations.find((violation) => violation.id === row.id);
    if (currentViolation && currentViolation.violations.length > 0)
      return currentViolation.violations.map((violation: any) => (
        <div className={styles.subRow}>
          {this.subHeaders.map((cell) => (
            <div key={cell.key} className={`${styles.tableCell} ${styles[cell.key]}`}>
              {this.renderDetail(cell.key, violation[cell.key])}
            </div>
          ))}
        </div>
      ));
    else {
      return (
        <div className={styles.subRow}>
          {this.subHeaders.map((cell) => (
            <div key={cell.key} className={`${styles.tableCell} ${styles[cell.key]}`}>
              {this.renderDetail(cell.key, "---")}
            </div>
          ))}
        </div>
      );
    }
  };

  renderDetail = (key: any, value: any) => {
    switch (key) {
      case "metric":
        return <p className={styles.detailMetrics}>{value}</p>;
      case "message":
        return <p className={styles.detailMessage}>{value}</p>;
      default:
        return <p className={styles.tableTextarea}>{value || "---"}</p>;
    }
  };

  handleRowClick = (row: any) => {
    this.setState({
      isModalOpen: true,
      selectedPolicyIndex: row,
    });
  };

  render() {
    const { violations } = this.props;
    const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;
    const selectedViolation = violations[this.state.selectedPolicyIndex ?? -1];
    return (
      <>
        <ComposedModal
          isOpen={this.state.isModalOpen}
          composedModalProps={{ shouldCloseOnOverlayClick: true }}
          onCloseModal={() => this.setState({ isModalOpen: false })}
          modalHeaderProps={{ title: selectedViolation?.policyName }}
        >
          {() => (
            <ModalForm>
              <ModalBody>
                <h2 className={styles.modalSectionTitle}>{`Violations (${
                  selectedViolation?.violations?.length ?? 0
                })`}</h2>
                <ul className={styles.modalLabels}>
                  {(selectedViolation?.violations ?? [])?.map(({ metric, message }: any, index: any) => (
                    <li key={index}>
                      <span>{metric} : </span>
                      <span style={{ fontStyle: "italic" }}>{message}</span>
                    </li>
                  ))}
                </ul>
                <h2 className={styles.modalSectionTitle}>Failed Definition Types</h2>
                <p>
                  {selectedViolation?.policyDefinitionTypes && selectedViolation?.policyDefinitionTypes.length
                    ? selectedViolation?.policyDefinitionTypes.join(", ")
                    : "---"}
                </p>
                <h2 className={styles.modalSectionTitle}>Labels</h2>
                <ul className={styles.modalLabels}>
                  {Object.entries(selectedViolation?.labels ?? {})?.map((entry) => (
                    <li key={entry[0]}>
                      <Tag type="purple">
                        {entry[0]}
                        <span>:</span>
                        {entry[1]}
                      </Tag>
                    </li>
                  ))}
                </ul>
                <h2 className={styles.modalSectionTitle}>Reference Link</h2>
                {selectedViolation?.referenceLink ? (
                  <a href={selectedViolation.referenceLink}>{selectedViolation.referenceLink}</a>
                ) : (
                  <p>No reference link provided</p>
                )}

                <h2 className={styles.modalSectionTitle}>Reference ID</h2>
                <p>{selectedViolation?.referenceId ?? "---"}</p>
                <h2 className={styles.modalSectionTitle}>Created Date</h2>
                <time>
                  {selectedViolation?.policyActivityCreatedDate
                    ? moment(selectedViolation.policyActivityCreatedDate).format("MMMM DD, YYYY")
                    : "---"}
                </time>
                <h2 className={styles.modalSectionTitle}>Annotations</h2>
                {selectedViolation?.annotations ? (
                  <CodeSnippet
                    copyButtonDescription="Copy annotations to clipboard"
                    onClick={() => copy(JSON.stringify([selectedViolation?.annotations]))}
                    type="multi"
                  >
                    {JSON.stringify([selectedViolation?.annotations], null, 1)}
                  </CodeSnippet>
                ) : (
                  "---"
                )}
              </ModalBody>
              <ModalFooter>
                <Button onClick={() => this.setState({ isModalOpen: false })} data-testid="close-violation-detail">
                  Close
                </Button>
              </ModalFooter>
            </ModalForm>
          )}
        </ComposedModal>
        <DataTable
          rows={violations}
          headers={this.headers}
          isSortable={true}
          render={({ rows, headers, getHeaderProps, getRowProps }: any) => (
            <TableContainer>
              <Table isSortable className={styles.tableContainer} useZebraStyles={false}>
                <TableHead>
                  <TableRow className={styles.tableHeadRow}>
                    {headers.map((header: any) => (
                      <TableHeader
                        key={header.key}
                        {...getHeaderProps({ header, className: `${styles.tableHeader} ${styles[header.key]}` })}
                      >
                        {header.header}
                      </TableHeader>
                    ))}
                  </TableRow>
                </TableHead>
                <TableBody className={styles.tableBody}>
                  {rows.map((row: any, rowIndex: any) => (
                    <TableRow
                      className={styles.tableBodyRow}
                      key={row.id}
                      {...getRowProps({ row })}
                      onClick={() => this.handleRowClick(rowIndex)}
                      onKeyDown={(e: any) => isAccessibleEvent(e) && this.handleRowClick(rowIndex)}
                      tabIndex={0}
                      data-testid="violations-table-row"
                    >
                      {row.cells.map((cell: any, cellIndex: any) => (
                        <TableCell key={cellIndex} style={{ padding: "0" }}>
                          <div className={styles.tableCell}>{this.renderCell(rowIndex, cellIndex, cell.value)}</div>
                        </TableCell>
                      ))}
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            </TableContainer>
          )}
        />
      </>
    );
  }
}

export default ViolationsTable;
