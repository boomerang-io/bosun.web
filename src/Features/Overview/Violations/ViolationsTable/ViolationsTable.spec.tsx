import React from "react";
import ViolationsTable from "./index";
import { violations as violationsData } from "ApiServer/fixtures";

const violations = [violationsData[0]];
describe("ViolationsTable --- Snapshot", () => {
  it("Capturing Snapshot of ViolationsTable", async () => {
    const { baseElement } = (global as any).rtlRender(<ViolationsTable violations={violations} />);
    expect(baseElement).toMatchSnapshot();
  });
});
