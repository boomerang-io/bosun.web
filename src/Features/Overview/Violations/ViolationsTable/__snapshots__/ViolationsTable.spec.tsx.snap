// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViolationsTable --- Snapshot Capturing Snapshot of ViolationsTable 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="bx--data-table-container"
    >
      <div
        class="bx--data-table-content"
      >
        <table
          class="bx--data-table tableContainer bx--data-table--sort bx--data-table--no-border"
        >
          <thead>
            <tr
              class="tableHeadRow"
            >
              <th
                aria-sort="none"
                class="tableHeader policyName"
                scope="col"
              >
                <button
                  class="tableHeader policyName bx--table-sort"
                  type="button"
                >
                  <span
                    class="bx--table-sort__flex"
                  >
                    <div
                      class="bx--table-header-label"
                    >
                      Policy
                    </div>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="tableHeader nbrViolations"
                scope="col"
              >
                <button
                  class="tableHeader nbrViolations bx--table-sort"
                  type="button"
                >
                  <span
                    class="bx--table-sort__flex"
                  >
                    <div
                      class="bx--table-header-label"
                    >
                      Violations
                    </div>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="tableHeader policyDefinitionTypes"
                scope="col"
              >
                <button
                  class="tableHeader policyDefinitionTypes bx--table-sort"
                  type="button"
                >
                  <span
                    class="bx--table-sort__flex"
                  >
                    <div
                      class="bx--table-header-label"
                    >
                      Failed Definition Types
                    </div>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="tableHeader referenceId"
                scope="col"
              >
                <button
                  class="tableHeader referenceId bx--table-sort"
                  type="button"
                >
                  <span
                    class="bx--table-sort__flex"
                  >
                    <div
                      class="bx--table-header-label"
                    >
                      Reference ID
                    </div>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                  </span>
                </button>
              </th>
              <th
                aria-sort="none"
                class="tableHeader policyActivityCreatedDate"
                scope="col"
              >
                <button
                  class="tableHeader policyActivityCreatedDate bx--table-sort"
                  type="button"
                >
                  <span
                    class="bx--table-sort__flex"
                  >
                    <div
                      class="bx--table-header-label"
                    >
                      Activity Date
                    </div>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      fill="currentColor"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                      />
                    </svg>
                  </span>
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            aria-live="polite"
            class="tableBody"
          >
            <tr
              class="tableBodyRow"
              data-testid="violations-table-row"
              tabindex="0"
            >
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    Glens Policy
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    1
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    Static Code Analysis, Static Code Analysis
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    ---
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <time
                    class="tableTextarea"
                  >
                    January 01, 2020
                  </time>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
